### PR TITLE
fix: Fix to delete cookie when AppSubURL is non-empty

### PR DESCRIPTION
**Pull request by jtran**
# Problem

We were observing a redirect loop, either when logging out or when trying to view a private repo with an expired session. This only seems to happen when `AppSubURL` is non-empty.

Using `release/v1.21`, several commits after the 1.21.10 release, on a server recently upgraded from 1.20.

Steps to reproduce:

1. Visit https://mydomain.com/sub_path/user/login (where `ROOT_URL` is configured to `https://mydomain.com/sub_path/`)
2. Click Remember This Device
3. Login
4. In the browser's dev tools, alter the Path of the gitea cookie from `/sub_path` to `/sub_path/`
5. Attempt to logout
6. You will now be redirected back to the home page

The logout only clears cookies in the `/sub_path`, not `/sub_path/`. Due to PR https://github.com/go-gitea/gitea/pull/29599, the login page now sees the gitea cookie that can't be deleted, auto logs in, and then redirects back to the home page.

Existing code already tries to delete the cookie at `/sub_path`. https://github.com/go-gitea/gitea/pull/24107/files#diff-7c540b84d46e33f1e7b33c7a4cc4daed15b765c11da81070d2313b42251aa48eR48-R53. But it doesn't delete cookies at `/sub_path/`. https://github.com/go-gitea/gitea/pull/29552 changed the default value of the path, but it didn't update code that clears it.

This is problematic because the longer path is more specific as far as the browser is concerned, and the browser prefers the old, outdated cookie with the trailing slash over newly set cookies. A similar problem may occur when no path is set on the cookie since browsers use the current page's path to determine the path of the cookie. The new, correct cookie path at the root should be `/`, but cookies without a path set may override it.

The workaround was to clear browser cookies, which is not desirable.

# Solution

This PR changes it so that whenever a cookie is written, we also clear legacy cookies that could override the cookie currently being written. In this way, cookies are lazily upgraded. In the interim, legacy cookies are still sent from the browser to the server and used. This is fine.

Cookies are written in `modules/web/middleware/cookie.go`, but also deep inside the chi dependency. Rather than modifying the latter, we clear legacy cookies whenever regenerating a session. This is required to handle the case of an expired session that would otherwise cause a redirect loop.

Open to suggestions on a better approach.

### CODE REVIEW

**Comment by jtran** [comment]
I made a mistake. I will recreate the PR or re-open once fixes have been made.

**Review by wxiaoguang — DISMISSED** [review]

File: modules/web/middleware/cookie.go [hunk 1]
```diff
@@ -48,7 +48,23 @@ func SetSiteCookie(resp http.ResponseWriter, name, value string, maxAge int) {
 	if maxAge < 0 {
 		// There was a bug in "setting.SessionConfig.CookiePath" code, the old default value of it was empty "".
 		// So we have to delete the cookie on path="" again, because some old code leaves cookies on path="".
-		cookie.Path = strings.TrimSuffix(setting.SessionConfig.CookiePath, "/")
-		resp.Header().Add("Set-Cookie", cookie.String())
+		// The code was updated, but it behaves differently depending on the
+		// value of AppSubURL.  When AppSubURL is non-empty, the cookie with a
+		// trailing slash must be deleted.
+		withoutTrailingSlash := strings.TrimSuffix(setting.SessionConfig.CookiePath, "/")
+		withTrailingSlash := ensureSuffix(setting.SessionConfig.CookiePath, "/")
```
Comment (wxiaoguang) [inline modules/web/middleware/cookie.go]:
```suggestion
		withoutTrailingSlash := strings.TrimSuffix(setting.SessionConfig.CookiePath, "/")
		withTrailingSlash := withoutTrailingSlash + "/"
```

File: modules/web/middleware/cookie.go [hunk 2]
```diff
@@ -48,7 +48,23 @@ func SetSiteCookie(resp http.ResponseWriter, name, value string, maxAge int) {
...
+		for _, path := range []string{withoutTrailingSlash, withTrailingSlash} {
+			if setting.SessionConfig.CookiePath != path {
+				cookie.Path = path
+				resp.Header().Add("Set-Cookie", cookie.String())
+			}
```
Comment (wxiaoguang) [inline modules/web/middleware/cookie.go]:
```suggestion
		cookie.Path = path
		resp.Header().Add("Set-Cookie", cookie.String())
```

(the `resp.Header().Add("Set-Cookie", cookie.String())` before `if maxAge < 0 {` could be moved)

File: modules/web/middleware/cookie.go [hunk 3]
```diff
@@ -48,7 +48,23 @@ func SetSiteCookie(resp http.ResponseWriter, name, value string, maxAge int) {
...
+		}
 	}
 }
+
+func ensureSuffix(s, suffix string) string {
+	if strings.HasSuffix(s, suffix) {
+		return s
+	}
+	return s + suffix
+}
```
Comment (wxiaoguang) [inline modules/web/middleware/cookie.go]:
```suggestion
```

File: modules/web/middleware/cookie.go (hunk 2 above)
Comment (wxiaoguang) [reply modules/web/middleware/cookie.go]: I mean:

```
if maxAge < 0 {
   for ... {
      cookie.Path = path
      resp.Header().Add("Set-Cookie", cookie.String())
   }
} else {
  resp.Header().Add("Set-Cookie", cookie.String())
}
```

Comment (jtran) [reply modules/web/middleware/cookie.go]: I'm not sure how the line could be moved while also deleting the condition. `setting.SessionConfig.CookiePath` may or may not end with a trailing slash, depending on the value of `AppSubURL`. I updated it to not duplicate `resp.Header().Add("Set-Cookie", cookie.String())`, so hopefully this is what you mean.

File: modules/web/middleware/cookie.go [hunk 4]
```diff
@@ -38,17 +38,23 @@ func SetSiteCookie(resp http.ResponseWriter, name, value string, maxAge int) {
...
+	// of AppSubURL.  When AppSubURL is non-empty, the cookie with a trailing
+	// slash must be deleted.
+	withoutTrailingSlash := strings.TrimSuffix(setting.SessionConfig.CookiePath, "/")
+	withTrailingSlash := withoutTrailingSlash + "/"
+	for _, path := range []string{withoutTrailingSlash, withTrailingSlash} {
+		if maxAge < 0 || setting.SessionConfig.CookiePath == path {
+			cookie.Path = path
+			resp.Header().Add("Set-Cookie", cookie.String())
+		}
 	}
```
Comment (wxiaoguang) [inline modules/web/middleware/cookie.go]:
```suggestion
	if maxAge < 0 {
		withoutTrailingSlash := strings.TrimSuffix(setting.SessionConfig.CookiePath, "/")
		withTrailingSlash := withoutTrailingSlash + "/"
		for _, path := range []string{withoutTrailingSlash, withTrailingSlash} {
			cookie.Path = path
			resp.Header().Add("Set-Cookie", cookie.String())
		}
	} else {
		cookie.Path = setting.SessionConfig.CookiePath
		resp.Header().Add("Set-Cookie", cookie.String())
	}
```

**Review by lunny — APPROVED** [review]

**Comment by jtran** [comment]
Found more browsers that hadn't been cleared, and after some more testing, it seems like this doesn't completely solve the problem. Converted the PR to draft until we figure it out.

**Comment by kdumontnu** [comment]
I think we need to settle on what the desired handling of cookies is, because it's not clear to me.

Here is what I propose (please correct me if anyone has other insights):
- The application SHOULD treat cookies with trailing `/` equivalent to no trailing slash (i.e. ignore trailing slash)
  - This is mostly needed to fix the intermediate issue where we users have both cookies with and without trailing slash
- The application should NOT delete or modify cookies at the root domain if there is a sub path (i.e. always use AppPath for cookies)
  - Because users could have installations at `domain/path1` and `domain/path2`

Is this correct?

I think if a user as a cookie at the root domain, but the application has a path, the cookie will be ignored.

**Comment by silverwind** [comment]
Based on https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#path_attribute, I recommend:

- if the parsed `path` of ROOT_URL is either `"/"` or `""`, set cookie on `"/"`
- if the parsed `path` of ROOT_URL is any other value, strip a single trailing slash off it and set cookie `path` to the resulting value.

**Comment by jtran** [comment]
> Based on reading [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#path_default_value), I recommend:
> 
>   * if the parsed `path` of `ROOT_URL` is either `"/"` or `""`, set cookie `path` to `"/"`
>   * For any other value, strip a single trailing slash off it and set cookie `path` to the resulting value.

#29534 should already do that.

The issue is that there are cookies leftover from the old approach in people's browsers.

I think this PR fixes logging out, for users who were already logged in. But if someone is logged out and trying to login, code will still try to use the old cookie with the incorrect path.

I'm working on an additional change that clears cookies when logging in, just before setting them again.

**Comment by silverwind** [comment]
> The issue is that there are cookies leftover from the old approach in people's browsers.

Sounds like a good idea to issue `path=$oldpath; expires=Thu, 01 Jan 1970 00:00:00 GMT` during logout or login to eliminate any previously wrongly set cookies.

**Comment by silverwind** [comment]
I think instead of doing this in the middleware, it'd be better to make `setting.SessionConfig.CookiePath` have the correct value in first place.

**Comment by kdumontnu** [comment]
@wxiaoguang any other insight?

File: modules/web/middleware/cookie.go [hunk 5]
```diff
@@ -45,10 +45,29 @@ func SetSiteCookie(resp http.ResponseWriter, name, value string, maxAge int) {
...
 	resp.Header().Add("Set-Cookie", cookie.String())
-	if maxAge < 0 {
-		// There was a bug in "setting.SessionConfig.CookiePath" code, the old default value of it was empty "".
-		// So we have to delete the cookie on path="" again, because some old code leaves cookies on path="".
-		cookie.Path = strings.TrimSuffix(setting.SessionConfig.CookiePath, "/")
-		resp.Header().Add("Set-Cookie", cookie.String())
+	// Previous versions would use a cookie path with a trailing /.
+	// These are more specific than cookies without a trailing /, so
+	// we need to delete these if they exist.
+	DeleteSiteCookieWithTrailingSlash(resp, name)
```
Comment (wxiaoguang) [inline modules/web/middleware/cookie.go]: If there was a cookie left on `path=""`, and now `CookiePath="/"`, should the old cookie on  `path=""` be deleted?

File: services/auth/source/oauth2/store.go [hunk 6]
```diff
@@ -65,6 +67,7 @@ func (st *SessionsStore) Save(r *http.Request, w http.ResponseWriter, session *s
 	chiStore := chiSession.GetSession(r)
 
 	if session.IsNew {
+		middleware.DeleteSiteCookieWithTrailingSlash(w, setting.SessionConfig.CookieName)
 		_, _ = chiSession.RegenerateSession(w, r)
```
Comment (wxiaoguang) [inline services/auth/source/oauth2/store.go]: Maybe the call to `DeleteSiteCookieWithTrailingSlash` could be put in our `session.RegenerateSession`, and change this `chiSession.RegenerateSession` to our `session.RegenerateSession`?

Comment (jtran) [reply services/auth/source/oauth2/store.go]: I made this change. Hopefully there isn't a reason to call `chiSession` directly.

**Review by wxiaoguang — APPROVED** [review]
Let's take a try

**Comment by jtran** [comment]
I broke some integration tests, so I'm working on that.

**Comment by wxiaoguang** [comment]
TBH I am not sure whether my question about cookie path `""` vs `"/"` is right, whether they are the same in browsers ....

**Comment by jtran** [comment]
> TBH I am not sure whether my question about cookie path `""` vs `"/"` is right, whether they are the same in browsers ....
> 
> If it is unclear/risky, maybe we should skip the `""` vs `"/"` case .... just like your previous approach.

They're not the same. When the Go code sees a path of `""`, it doesn't set a path on the cookie. The [default path](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#path_default_value) when a path is not set depends on the URI of the page when it's set. I don't think we want cookies to be set for various paths depending on the page that was accessed. Therefore, I'm pretty sure that the latest approach of clearing out the cookie with path `""` in favor of `"/"` is correct. Preferring `"/"` is much simpler to reason about. There will only be one at the root, not potentially many overriding each other depending on the current URI of the page.

**Comment by jtran** [comment]
Reverting the change to delete cookies with no path fixed the integration test failures. So maybe we don't want that change. I'm a little unclear about the reasoning behind this.

**Review by silverwind — APPROVED** [review]

**Pull request merged**